### ISSUE TITLE
CA1822 Is Overly Aggressive

### ISSUE THREAD

**Issue by david-maw** [dotnet/roslyn-analyzers#7577]
Warning CA1822 Is “‘xxx’ does not access instance data and can be marked as static”, the problem is that if (as in my case) xxx happens to be a public property used by XAML marking it as static will break the connection to XAML and indeed the documentation for CA1822 points out that if xxx is public marking it static may be a breaking change.

It would be nice if I could tell it to ignore this warning for public properties, but I could not figure out how, so I set out to suppress the property in the source. The trouble is, doing that generates IDE0079 “Remove unnecessary suppression” and it’s not clear to me what’s unnecessary about this suppression.

This only becomes more challenging when you are dealing with a large body of code (I was refactoring 14,000 lines) and there are many such warnings, some actionable, some not.

There's a sample MAUI project using MVVM to illustrate the problem at https://github.com/david-maw/Warning.git, run it and tap the button to change the message.

The message comes from a string property which reads a static string, the button command invokes OnPropertyChanged to notify the page to update the displayed string.

The property definition is easy enough to change to
``` C#
    public static string Msg => initialMsg ? "Hello, World!" : "Goodbye, World!";
```
The corresponding change to the XAML is trickier, because you must find it, but something like this works to display a message using a label control:
``` XAML
   Text="{x:Static vm:MainViewModel.Msg}"
```
Unfortunately, clicking the button will no longer change the text because `OnPropertyChanged` no longer notifies the page.

The only workaround I've found is to disable the warning globally, but that seems a bit drastic.

**Comment by LWChris** [comment]
It isn't pretty, but you can add a second `#pragma warning disable` for `IDE0079` around it:

```
#pragma warning disable IDE0079
#pragma warning disable CA1822
public string Msg => initialMsg ? "Hello, World!" : "Goodbye, World!";
#pragma warning restore CA1507
#pragma warning restore IDE0079
```

Note: unfortunately, disabling both in the same directive (`#pragma warning restore IDE0079, CA1507`) doesn't work, you will get two `IDE0079` on that very line. It seems to me that the code issuing `IDE0079` is quite buggy and doesn't reliably track whether the suppressed warning would be emitted or not.

**Comment by david-maw** [comment]
Thanks @LWChris that works fine as a workaround in a small code base but with many such warnings I found it  impractical to add everywhere it would be needed.

**Comment by LWChris** [comment]
Yeah. The other thing you can do for the time being is to disable the IDE0079 with a suppression file.

Or you can rewrite your property:

```
public string ThisWarns => SomeConstant;

public string ThisDoesnt { get; } = SomeConstant;
```

This only works for constant expressions, but since your property does not implement `INotifyPropertyChanged` (or else the warning would go away anyway) the value is probably quasi-constant from XAML's perspective.

A third option is to initialize the value from the ctor.

**Comment by david-maw** [comment]
@LWChris I may be misunderstanding you, the test ViewModel I provided looks like this

``` C#
internal partial class MainViewModel : INotifyPropertyChanged
{
    public MainViewModel() => ChangeMessageCommand = 
         new Command(() => { initialMsg = !initialMsg; OnPropertyChanged(nameof(Msg)); });

    private static bool initialMsg = true;
    public ICommand ChangeMessageCommand => field;
    public string Msg => initialMsg ? "Hello, World!" : "Goodbye, World!";
    private void OnPropertyChanged(string propertyName) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    public event PropertyChangedEventHandler? PropertyChanged;
}
```
Invoking `ChangeMsgCommand` inverts `initialMsg` and triggers a call on `OnPropertyChanged(nameof(Msg)`. `Msg` returns a different constant depending on the value of `InitialMsg`.

I'm not sure what you mean by "your property does not implement `NotifyPropertyChanged`". The containing class does implement `INotifyPropertyChanged`". Do you mean that the property does not have a setter calling `OnPropertyChanged`?

A key issue here seems to be that `initialMsg` is static, if it were not the warning would go away. I'm pretty sure this is not how I would implement this today but the code in question is old and non-constant static data has all sorts of pitfalls. Nevertheless, this is legal so maybe a fix is to limit CA1822 to private declarations and add an alternative for `public` and `internal` ones?

**Comment by LWChris** [comment]
> [...] I'm pretty sure this is not how I would implement this today but the code in question is old [...] Nevertheless, this is legal [...]

"Limit CA1822 to private members" is not a sensible way of fixing this kind of false positive. The code is legal, but uncommon. I think analyzers should not attempt to cater _legal_ code by sacrificing usefulness for _common_ code.

This is exactly what suppressions are meant for, so I'd rather have #7554 fixed than CA1822 changed.

**Comment by david-maw** [comment]
@LWChris, I agree that simply limiting CA1822 to private (and probably internal) members without introducing an analogous warning for public ones would be problematic, that's why I suggested introducing a new warning as well.

In general, it seems tricky to make assertions about public members as you've no idea how they might be used in XAML or otherwise and recommendations like this one that make the code continue to compile but stop actually working make me uncomfortable.

That said, I agree fixing #7554 would be a Good Thing.

**Comment by LWChris** [comment]
But how would a hypothetical `CA1822_b` for public members be any different from the current CA1822?

It would still look at the same criteria. It would still flag the same members. It would still have to say in the resolution advisory that it may be a false positive that should be suppressed rather than acted upon. That what CA1822 already does.
